### ISSUE TITLE
Empty unions cause a crash

### ISSUE THREAD

**Issue by joneshf** [dhall-lang/dhall-haskell#546]
Looks like, if we try to use an empty union in a lambda, the program crashes.

```
$ dhall
\(void : <>) -> 0
<stderr>: hPutChar: invalid argument (invalid character)
```

i think the spec allows empty union types, so maybe it's something specific to the implementation? Any thoughts before I dive in looking for the problem?

**Comment by sellout** [comment]
This works for me in 1.15.1. I’m assuming you’re on 1.16.x? Or master?

It looks like it’s erroring when trying to pretty-print the type to stderr.

Can you `dhall format` it successfully? Have you used `<>` in other contexts successfully?

Or … is the LOCALE set “incorrectly” or something? This does look like the errors we used to get in CI, before Dhall was fixed to ignore the LOCALE.

**Comment by joneshf** [comment]
Oh, you're right, it's the locale thing!

The particular line it's trying to print is a red herring. It's failing whenever it tries to display unicode:

```
$ dhall
\(x : Text) -> x
<stderr>: hPutChar: invalid argument (invalid character)
```

I was running 1.15.1 in a pure nix shell, so it wasn't loading the environment variables for locale stuff:

```
$ nix-shell --packages dhall --pure -I nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/6afd19e699c1505938d54312b522562fd75c8062.tar.gz      

[nix-shell:~]$ dhall version
1.15.1

[nix-shell:~]$ dhall 
Natural/show
Natural <stderr>: hPutChar: invalid argument (invalid character)

[nix-shell:~]$ dhall format
Natural/show
Natural/show
[nix-shell:~]$ locale
LANG=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
```

Setting the locale seems to work:

```
[nix-shell:~]$ export LANG=en_US.UTF-8

[nix-shell:~]$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=

[nix-shell:~]$ dhall
Natural/show
Natural → Text

Natural/show
```

Sorry for the noise!

**Comment by joneshf** [comment]
Sorry, I closed this since #506 seems to address it:

```
$ nix-shell --pure shell-with-dhall-1.16.1.nix 

[nix-shell:~]$ dhall version
1.16.1

[nix-shell:~]$ dhall
\(void : <>) -> 0
∀(void : <>) → Natural

λ(void : <>) → 0

[nix-shell:~]$ locale
LANG=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=
```
